python/rmm/_lib/memory_resource.pyx

```diff
@@ -0,0 +1,288 @@
+    size_base : int, optional
+        The base of allocation block sizes (defaults is 2).
+    min_size_exponent : int, optional
+        The exponent of the minimum fixed block size to allocate
+        (default is 18).
+    max_size_exponent : int, optional
+        The exponent of the maximum fixed block size to allocate
+        (default is 22).
+    initial_blocks_per_size : int, optional
+        The number of blocks to preallocate from the upstream memory resource,
+        and to allocate when all current blocks are in use.
+    """
+    def __cinit__(
+        self,
+        MemoryResource upstream,
+        size_t size_base=2,
+        size_t min_size_exponent=18,
+        size_t max_size_exponent=22,
+        size_t initial_blocks_per_size=128
+    ):
+        self.c_obj.reset(new fixed_multisize_memory_resource_wrapper(
+            upstream.c_obj,
+            size_base,
+            min_size_exponent,
+            max_size_exponent,
+            initial_blocks_per_size
+        ))
+
+
+cdef class HybridMemoryResource(MemoryResource):
+    """"
+    Allocates memory from one of two allocators based on the requested size.
+
+    Parameters
+    ----------
+    small_alloc_mr : MemoryResource
+        The memory resource to use for small allocations.
+    large_alloc_mr : MemoryResource
+        The memory resource to use for large allocations.
+    threshold_size : int, optional
+        Size in bytes representing the threshold beyond which `large_alloc_mr`
+        is used.
+    """
+    def __cinit__(
+        self,
+        MemoryResource small_alloc_mr,
+        MemoryResource large_alloc_mr,
+        size_t threshold_size=1<<22
+    ):
+        self.c_obj.reset(
+            new hybrid_memory_resource_wrapper(
+                small_alloc_mr.c_obj,
+                large_alloc_mr.c_obj,
+                threshold_size
+            )
+        )
+
+
+cdef class LoggingResourceAdaptor(MemoryResource):
+    """
+    Memory resource that logs information about allocations/deallocations
+    performed by an upstream memory resource.
+
+    Parameters
+    ----------
+    upstream : MemoryResource
+        The upstream memory resource.
+    log_file_name : str
+        Path to the file to which logs are written.
+    """
+    def __cinit__(self, MemoryResource upstream, object log_file_name=None):
+        if log_file_name is None:
+            log_file_name = os.getenv("RMM_LOG_FILE")
+            if not log_file_name:
+                raise TypeError(
+                    "RMM log file must be specified either using "
+                    "log_file_name= argument or RMM_LOG_FILE "
+                    "environment variable"
+                )
+        self.c_obj.reset(new logging_resource_adaptor_wrapper(
+            upstream.c_obj,
+            log_file_name
+        ))
+
+    def flush(self):
+        (<logging_resource_adaptor_wrapper*>(self.c_obj.get()))[0].flush()
```

Comment:
Do we want to consider making this a `cpdef` so it can be more efficiently called from within Cython code?

Reply:
Made it a `cpdef`